Future / proposed spec changes

bamurtaugh opened the pull request:
Updating based on https://github.com/devcontainers/cli/pull/228 and https://github.com/devcontainers/spec/pull/89.

As discussed, I believe we want to wait to merge these changes for another few weeks, but I wanted to open the PR for discussion now.

Commit:
Add parallel execution

Commit:
Add feature env var

Chuxel commented on _implementors/spec.md:
```diff
@@ -194,6 +194,27 @@ To resume the environment from a stopped state:
 
 Like during the create process, remote [environment variables](#environment-variables) and [user](#users) configuration should be applied to all created processes in the container (inclusive of `userEnvProbe`).
 
+## <a href="#parallel-exec" name="parallel-exec" class="anchor"> Parallel lifecycle script execution (proposal) </a>
```
I think this is another one like labels that we can probably just call "implemented". We kept the proposal around for features and templates because they were so complicated, and there was a bunch of active conversation yet.

bamurtaugh replied:
Makes sense, removed!

Chuxel commented on _implementors/spec.md:
````diff
@@ -194,6 +194,27 @@ To resume the environment from a stopped state:
+}
+```
+
+You may review this proposal directly in [the spec](https://github.com/devcontainers/spec/blob/main/proposals/parallel-lifecycle-script-execution.md).
````
Given the content above is the spec, we can probably just omit this. 

```suggestion
You may review this proposal directly in [the spec](https://github.com/devcontainers/spec/blob/main/proposals/parallel-lifecycle-script-execution.md).
```

bamurtaugh replied:
Removed!

Chuxel reviewed (commented):
Yeah this makes sense in these spots.  👍

For parallel scripts though, this also means that the  json_reference needs a tweak since type for the "Lifecycle scripts" commands in the is now string, array, or object.  There's a description of string vs array there, so I think we'd need to summarize parallel execution there too.

Like in the other case, probably makes sense to add this to the respective spots in devcontainers/spec so we don't have a snafu on copying across later. :smile:

Few other comments below!

Commit:
Remove proposal word

Commit:
Removal

Commit:
Add info on object format

Commit:
Add id proposal

bamurtaugh committed:
Merge branch 'gh-pages' into bamurtaugh/other-spec

bamurtaugh commented:
Thanks for the review @Chuxel! I've done the following:

- Updated lifecycle scripts info to reference object type, added info on object type to the existing section on array vs string
- Pulled in changes from https://github.com/devcontainers/spec/issues/62
- Open https://github.com/devcontainers/spec/pull/128 for associated spec changes

Chuxel commented on _implementors/features.md:
```diff
@@ -311,3 +321,64 @@ There are several things to keep in mind for an application that implements Feat
 - Each Feature script executes as its own layer to aid in caching and rebuilding.
 - There is also a proposal to move properties like `privileged` to `devcontainer.json` for consistency and to reduce reliance on the `runArgs` property. See https://github.com/devcontainers/spec/issues/2.
+
+## <a href="#dev-container-id" name="dev-container-id" class="anchor"> Dev Container ID </a>
```
Feels like this belongs as a sub-heading under "devcontainer-feature.json properties".

It's also supported in normal devcontainer.json, correct @chrmarti?

Assuming so, we should also include `"${devcontainerId}"` in the variables section of json_reference.md (along with the properties it applies to... see next note).

bamurtaugh replied:
Moved underneath "devcontainer-feature.json properties" and added `${devcontainerId}` to the variables table!

Chuxel commented on _implementors/features.md:
```diff
@@ -311,3 +321,64 @@ There are several things to keep in mind for an application that implements Feat
+
+## <a href="#dev-container-id" name="dev-container-id" class="anchor"> Dev Container ID </a>
+
+### <a href="#dev-container-id-goal" name="dev-container-id-goal" class="anchor"> Goal </a>
+
+Allow Features to refer to an identifier that is unique to the dev container they are installed into and that is stable across rebuilds.
+
+E.g., the `docker-in-docker` Feature needs a way to mount a volume per dev container.
+
+### <a href="#dev-container-id-prop" name="dev-container-id-prop" class="anchor"> Proposal </a>
```
We can probably drop this being in a proposal and consider it part of the features implementation at this point given its merged into the CLI and will land in extensions soon.

```suggestion
```

Chuxel commented on _implementors/features.md:
```diff
@@ -311,3 +321,64 @@ There are several things to keep in mind for an application that implements Feat
+
+### <a href="#dev-container-id-prop" name="dev-container-id-prop" class="anchor"> Proposal </a>
+
+The identifier will be referred to as `${devcontainerId}` in the devcontainer.json and the Feature metadata and that will be replaced with the dev container's id. It should only be used in parts of the configuration and metadata that is not used for building the image because that would otherwise prevent pre-building the image at a time when the dev container's id is not known yet. Excluding bool ...
```
Most of these properties are for devcontainer.json right now. There's only three for features: `entrypoint`, `mounts`, `customizations` 

The rest would make sense in the json_reference table for variables.

bamurtaugh replied:
> The rest would make sense in the json_reference table for variables.

Could you give an example of what you're envisioning? I added an entry for `${devcontainerId}` in the variables table - are you envisioning we'd list the above devcontainer.json and Features properties within the `${devcontainerId}` entry in this table?

Chuxel replied:
Yeah this list is for devcontainer.json, not features:
`name`, `runArgs`, `initializeCommand`, `onCreateCommand`, `updateContentCommand`, `postCreateCommand`, `postStartCommand`, `postAttachCommand`, `workspaceFolder`, `workspaceMount`, `mounts`, `containerEnv`, `remoteEnv`, `containerUser`, `remoteUser`, `customizations`

But this probably qualifies as `any` since it's any string property... the others are enum, numbers or booleans.

Similarly, the only feature properties are `entrypoint`, `mounts`, `customizations`. I suspect `containerEnv` also works and that was an accidently omitted. 

Correct @chrmarti ?  

For features, we could just say "any string property" in both cases. That would future proof us as other htings like onCreateCommand are added to features.

bamurtaugh replied:
I removed the non-Feature properties from this list in features.md, and I added the list of devcontainer.json properties in the variable table list in the json_reference. Let me know if this is what you had in mind.

Chuxel reviewed (commented):
A few thoughts (which would also apply to https://github.com/devcontainers/spec/pull/128)

bamurtaugh committed:
Update _implementors/features.md

Co-authored-by: Chuck Lantz <clantz@microsoft.com>

Commit:
Make sub-heading

Commit:
Add to variable table

bamurtaugh committed:
Merge branch 'gh-pages' into bamurtaugh/other-spec

Commit:
Feedback

bamurtaugh commented:
Thanks so much for the great review and collab @Chuxel. I'll merge this PR and https://github.com/devcontainers/spec/pull/128 after the next Dev Containers release next week.

chrmarti commented on _implementors/features.md:
```diff
@@ -310,4 +373,4 @@ There are several things to keep in mind for an application that implements Feat
 - `containerEnv` is added before the feature is executed as `ENV` commands in the Dockerfile.
 - Each Feature script executes as its own layer to aid in caching and rebuilding.
-- There is also a proposal to move properties like `privileged` to `devcontainer.json` for consistency and to reduce reliance on the `runArgs` property. See https://github.com/devcontainers/spec/issues/2.
+- There is also a proposal to move properties like `privileged` to `devcontainer.json` for consistency and to reduce reliance on the `runArgs` property. See https://github.com/devcontainers/spec/issues/2.
```
We have added all but `entrypoint` (which needs more consideration). (We added `mounts`, `containerEnv`, `containerUser`, `init`, `privileged`, `capAdd`, and `securityOpt`.)

Chuxel replied:
Good point - we could just remove that line at this point.

bamurtaugh replied:
Thanks! Removed this line from this PR and the spec PR.

chrmarti reviewed (approved):
LGTM, left one comment on the additional devcontainer.json properties.

bamurtaugh committed:
Update features.md

bamurtaugh commented:
As the release is out, merging now. Thanks!
